### PR TITLE
refactor: add natspec documentation

### PR DESCRIPTION
**Pull request by MaxMustermann2**
- Where the word `operator` was used exclusively in the context of Bootstrap, replace it with `validator`
- Add natspec documentation and License for each file in `src/core`, `src/interfaces` (but not `src/interfaces/precompiles`) and `src/storage`
- Combine some duplicated code for add/register whitelist tokens in `ExocoreGateway`
- Remove unused functions / errors / events

## Summary by CodeRabbit

- **New Features**
  - Introduced a new interface for managing validators, enhancing the application's capability to register and manage validators effectively.
  - Added new events for better tracking of contract activities, including `RestakingActivated` and `WithdrawalSuccess`.

- **Improvements**
  - Updated naming conventions and functionality to shift focus from operators to validators, clarifying the contract's purpose.
  - Enhanced documentation within contracts for better clarity regarding functionalities and improved error handling.
  - Added modifiers and checks to improve security and control over function access.
  - Clarified the responsibilities of the `ClientChainGateway` and structured user fund management in the `Vault`.

- **Bug Fixes**
  - Removed outdated data structures related to user balance management, simplifying the process.

- **Documentation**
  - Updated related documentation to reflect changes in functionality and structure related to validators, enhancing overall usability.

### CODE REVIEW

File: src/core/ExocoreGateway.sol [hunk 1]
```diff
@@ -202,11 +183,37 @@ contract ExocoreGateway is
         string[] calldata names,
         string[] calldata metaData
     ) external onlyOwner whenNotPaused {
+        _addOrUpdateWhitelistTokens(clientChainId, tokens, decimals, tvlLimits, names, metaData, false);
```
Comment (adu-web3) [inline src/core/ExocoreGateway.sol:186]: nice refactoring, this looks cleaner

**Review by adu-web3 — APPROVED** [review]
good job! look good to me

File: docs/client-chain-contracts-design.md [hunk 2]
```diff
@@ -244,21 +244,7 @@ Exocore validator set should be the owner of `Controller` so that it can update
...
-    
-    // @notice this info is used to update specific user's owned tokens balance
-    struct UserBalanceUpdateInfo {
-        address user;
-        uint256 updatedAt;
-        TokenBalanceUpdateInfo[] tokenInfo;
-    }
-    
+
     event DepositResult(address indexed depositor, bool indexed success, uint256 amount);
```
Comment (bwhour) [inline docs/client-chain-contracts-design.md:248]: this part need update.  the event results has been updated  formate in exocoregateway.

Comment (MaxMustermann2) [reply docs/client-chain-contracts-design.md]: Updated the documentation; left out events to focus on the client chain contracts

File: docs/client-chain-contracts-design.md [hunk 3]
```diff
@@ -1,32 +1,32 @@
...
 
 We have these components included in Exocore client chain smart contracts architecture:
 
-1. `Gateway`: This is the entry point where client chain users make request to Exocore validator set, as well as the end point that receives cross-chain messages from Exocore validator set.
-2. `Vault`: This is where user funds are taken into custody and managed. Within `Vault`, user balance is updated periodically by Exocore validator set through cross-chain message to reveal user’s real position(after slashing, rewarding and other impact). Users can withdraw from `Vault` based on grant from Exocore validator set. Every specific asset should have standalone `Vault`.
-3. `Controller`: The controller that is responsible for managing multiple `Vault`s. It should be the entry point for operations on `Vault`, as well as the entry point for user’s interactions with Exocore validator set.
+1. `ClientChainGateway`: This is the entry point where client chain users make request to Exocore validator set, as well as the end point that receives cross-chain messages from Exocore validator set.
```
Comment (bwhour) [inline docs/client-chain-contracts-design.md:14]: Exocore validator set : this address has been removed,  and now this is the owner address. so all the document related to this should be updated.

Comment (MaxMustermann2) [reply docs/client-chain-contracts-design.md:14]: Removed in all places except 2, where it talks about the validator set eventually owning the contract.

File: src/storage/BootstrapStorage.sol [hunk 4]
```diff
@@ -421,8 +328,13 @@ contract BootstrapStorage is GatewayStorage {
...
     }
 
-    function isValidExocoreAddress(string calldata operatorExocoreAddress) public pure returns (bool) {
-        bytes memory stringBytes = bytes(operatorExocoreAddress);
+    /// @notice Checks if the provided Exocore address is valid.
+    /// @param question The Exocore address to check.
+    /// @return True if the address is valid, false otherwise.
+    /// @dev Since implementation of bech32 is difficult in Solidity, this function only
+    /// checks that the address is 42 characters long and starts with "exo1".
+    function isValidExocoreAddress(string calldata question) public pure returns (bool) {
```
[the remark this answers is not in the document]
Comment (MaxMustermann2) [reply src/storage/BootstrapStorage.sol:336]: No. We don't know if it is an `exoAddress`, so how can we call it that?

**Review by bwhour — APPROVED** [review]

**Pull request merged**